src/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs

```diff
@@ -614,10 +616,100 @@ private static void AddObjectsIntoTargetCollection(object targetCollection, List
             }
         }
 
-        private static void SetMemberValue(object o, object value, string memberName)
+        private static IDictionary<Tuple<Type, Type>, Func<MemberInfo, SetMemberValueDelegate>> s_GetSetMemberValueDelegateWithTypeCache = new ConcurrentDictionary<Tuple<Type, Type>, Func<MemberInfo, SetMemberValueDelegate>>();
+        private static IDictionary<Tuple<Type, string>, SetMemberValueDelegate> s_SetMemberValueDelegateCache = new ConcurrentDictionary<Tuple<Type, string>, SetMemberValueDelegate>();
+        private delegate void SetMemberValueDelegate(object o, object val);
+
+        private static SetMemberValueDelegate GetSetMemberValueDelegate(object o, string memberName)
         {
-            MemberInfo memberInfo = ReflectionXmlSerializationHelper.GetMember(o.GetType(), memberName);
-            SetMemberValue(o, value, memberInfo);
+            SetMemberValueDelegate result;
+            var typeMemberNameTuple = Tuple.Create(o.GetType(), memberName);
+            if (!s_SetMemberValueDelegateCache.TryGetValue(typeMemberNameTuple, out result))
+            {
+                MemberInfo memberInfo = ReflectionXmlSerializationHelper.GetMember(o.GetType(), memberName);
+                Type memberType;
+                if (memberInfo is PropertyInfo propInfo)
+                {
+                    memberType = propInfo.PropertyType;
+                }
+                else if (memberInfo is FieldInfo fieldInfo)
+                {
+                    memberType = fieldInfo.FieldType;
+                }
+                else
+                {
+                    throw new InvalidOperationException(SR.Format(SR.XmlInternalError));
+                }
+
+                var typeMemberTypeTuple = Tuple.Create(o.GetType(), memberType);
+                Func<MemberInfo, SetMemberValueDelegate> getSetMemberValueDelegateWithType;
+                if (!s_GetSetMemberValueDelegateWithTypeCache.TryGetValue(typeMemberTypeTuple, out getSetMemberValueDelegateWithType))
```

Comment:
We don't need it. It's main usage is if a class has multiple properties of the same type such as string. Would you prefer I remove it?

---
In reply to: [122582416](https://github.com/dotnet/corefx/pull/21155#discussion_r122582416) [](ancestors = 122582416)